sync.py

```diff
@@ -30,13 +35,183 @@ def __init__(self):
+                    old_hashes[vals[0]] = vals[1]
+
+        # Second, read in current values of data from Banner via WSAPI
+        # These are sorted twice in WSAPI: first by prog_code, and then sub-sorted by start_term_code
+        new_banner_data = json.loads(requests.get('https://wsapi.bethel.edu/program-data').content)
+        # Sourced from: https://stackoverflow.com/a/16839304
+        data_len_magnitude = int(math.log10(len(new_banner_data)))
+
+        def _leftpad_index(index, mag):
+            if index == 0:
+                # Can't take the log of 0
+                index_magnitude = 0
+            else:
+                # Take in an integer i, and left-pad zeroes until it has as many decimal places as specified by "mag"
+                index_magnitude = int(math.log10(index))
+            return '0' * (mag - index_magnitude) + str(index)
+
+        # Third, create a dictionary of new md5 hashes, each stored at a 3-part key:
+        # prog_code, start_term_code, and row index
+        new_hashes = {}
+        distinct_program_codes = set()
+        for i in range(len(new_banner_data)):
+            row = new_banner_data[i]
+            # Convert values from unicode to string
+            distinct_program_codes.add(str(row['prog_code']))
+            # Have to left-pad the index with zeroes so that sorted() retains proper index order
+            row_key = row['prog_code'] + '__' + row['start_term_code'] + '__' + _leftpad_index(i, data_len_magnitude)
+            new_hashes[row_key] = self.convert_dictionary_to_hash(row)
+
+        # This is only true on the first run, so the file has to be created
+        if not os.path.isfile(BANNER_HASHES_CSV_PATH):
+            with open(BANNER_HASHES_CSV_PATH, 'w+') as new_data_hashes:
+                for row_key in sorted(new_hashes.keys()):
+                    new_data_hashes.write('%s,\t%s,\t\n' % (row_key, new_hashes[row_key]))
+
+            # Return all rows as "new"
+            return new_banner_data, distinct_program_codes
+
+        # Fourth, check if any md5 hashes in new are different than in old, or if new has md5 hashes that old doesn't.
+        # If there are any differences, put the corresponding rows into an array that will eventually be returned.
+        # Due to how the rows are sorted in WSAPI and how the row_key is generated for the dictionaries, the indexes
+        # at the end of each row_key should be in order from 0-n. This is CRUCIAL to the sorting algorithm below!
+        old_index = 0
+        old_keys = sorted(old_hashes.keys())
+        new_index = 0
+        new_keys = sorted(new_hashes.keys())
+
+        # This variable is an offset to help compare new row indexes to old row indexes
+        # A positive value means there have been more rows added to new than rows deleted from old, and vice versa
+        num_rows_inserted_or_deleted = 0
+        different_or_new_rows = []
+
+        while old_index < len(old_keys) and new_index < len(new_keys):
```

Comment:
this whole while loop is way too complicated. Seems like there should be a much easier way to do this.

Reply:
Apologies in advance for a very long-winded answer.

I agree, this while loop is a **monster**. That being said, I'm not sure that it can be done more simply.

- It could be written less densely, but that doesn't simplify it. If anything, it would probably obfuscate how it actually does its work.
- It could be written with fewer lines, but then it would be even more dense. Instead of being able to read it and be appalled at how gross it is, people would have to think hard about what it's doing before becoming appalled.
- It could be written with more simple logic inside the while loop, but then the complicated logic would be moved somewhere else instead of just being removed like we would like.

To expound on "using more simple logic": technically, the only thing _needed_ for the hashes in the .csv is the index of the row returned by Banner, and then the logic would boil down to 
```
if old_hash[index] != new_hash[index]:
    diff_or_new_rows.append(new_data[index])
```
The drawback is that any time a row was inserted or deleted, then all rows after the inserted/deleted row would be marked as "new/different" due to an off-by-one error. Alternatively, reducing it from 3 checks (one if/elif/else for each of the 3 parts of the key) to just 1 check like so:
```
if old_key < new_key:
    # etc
elif old_key > new_key:
    # etc
else:
    if old_hash != new_hash:
        #etc 
```
This too, would logically work for catching changed rows, but then the off-by-one error from the indexes at the end of the keys would come back to haunt. With some fancy code gymnastics the offset variable could be implemented again, but then we would have just exchanged one form of code complexity (breaking the keys into their parts and doing granular comparisons on each part) for another (having to fetch, parse, offset, compare, and update the indexes of each key as the while loop grinds through the rows).

Another idea I considered initially would be to just have the old keys in a `Set`, the new keys in a different `Set`, then do some Union and Intersection math to get keys that were only in the "New" Set. But once more, the off-by-one error rears its ugly head. Any keys with different index suffixes would be marked as "not in the intersection". Obviously, we would just drop the index suffix and only keep program code/start term code keys in the Sets. This leads to the problem of "How do we know which rows were updated?" Again, it exchanges the current complexity for a different form of complexity elsewhere.

I hope that I've provided some concrete examples of how "it can have the complexity moved out of the while loop, but not removed entirely". I'm very open to questions and suggestions, though. The while loop is extremely distasteful and I would love if it could be refactored out of existence.

Reply:
Yeah i get that. I still think we can drastically simplify this, as we get to build both the old and new arrays.

My first thought, is to simply put the relevant data that is changing into a single string and md5 it. Then use dict keys to simple check if it exists. Ultimately, this is combining the data into unique keys. It essentially condenses the while loop into a simple loop to check.

I think the idea of using sets is good (maybe can use that in the idea i had above ^). 

I can think about this more, if you want.

Reply:
Take another look tomorrow morning; I think you'll be very pleased with the new version.

Reply:
thats about as close to a "best case" scenario as we can get. Nice job